Indexing/Index.hpp

```diff
@@ -19,102 +19,124 @@
+ * A LeafData must have a function  `<Key> key() const;` returns the key, and must have comparison operators (<=,<,>=,>,!=,==) implemented.
+ * See e.g. TermLiteralClause below for examples.
+ */
 namespace Indexing
 {
 using namespace Kernel;
 using namespace Lib;
 using namespace Saturation;
 
-
-/**
- * Class of objects which contain results of single literal queries.
- */
-template<class Unifier>
-struct LQueryRes
+struct LiteralClause 
 {
-  LQueryRes() {}
-  LQueryRes(Literal* l, Clause* c, Unifier unifier)
-  : literal(l), clause(c), unifier(std::move(unifier)) {}
+  Literal* const& key() const
+  { return literal; }
 
+private:
+  auto asTuple() const
+  { return std::make_tuple(clause->number(), literal->getId()); }
+public:
 
-  Literal* literal;
-  Clause* clause;
-  Unifier unifier;
+  IMPL_COMPARISONS_FROM_TUPLE(LiteralClause)
+
+  Literal* literal = nullptr;
+  Clause* clause = nullptr;
+
+  friend std::ostream& operator<<(std::ostream& out, LiteralClause const& self)
+  { return out << "{ " << outputPtr(self.clause) << ", " << outputPtr(self.literal) << " }"; }
+};
+
+template<class Value>
+struct TermWithValue {
+  TypedTermList term;
+  Value value;
+
+  TermWithValue() {}
+
+  TermWithValue(TypedTermList term, Value v)
+    : term(term)
+    , value(std::move(v))
+  {}
+
+  TypedTermList const& key() const { return term; }
+
+  auto asTuple() const
+  { return std::tie(term, value); }
+
+  IMPL_COMPARISONS_FROM_TUPLE(TermWithValue)
+
+  friend std::ostream& operator<<(std::ostream& out, TermWithValue const& self)
+  { return out << self.asTuple(); }
+};
+
+class TermWithoutValue : public TermWithValue<std::tuple<>> 
+{
+public:
+  TermWithoutValue(TypedTermList t) 
+    : TermWithValue(t, std::make_tuple()) 
+  { }
 };
 
-template<class Unifier>
-LQueryRes<Unifier> lQueryRes(Literal* l, Clause* c, Unifier unifier)
-{ return LQueryRes<Unifier>(l,c,std::move(unifier)); }
+struct TermLiteralClause 
+{
+  TypedTermList term;
+  Literal* literal = nullptr;
+  Clause* clause = nullptr;
+
+  TypedTermList const& key() const { return term; }
+
+  auto  asTuple() const
+  { return std::make_tuple(clause->number(), literal->getId(), term); }
+
+  IMPL_COMPARISONS_FROM_TUPLE(TermLiteralClause)
+
+  friend std::ostream& operator<<(std::ostream& out, TermLiteralClause const& self)
+  { return out << "("
+               << self.term << ", "
+               << self.literal
+               << outputPtr(self.clause)
+               << ")"; }
+};
 
 /**
  * Class of objects which contain results of term queries.
  */
-template<class Unifier>
-struct TQueryRes
+template<class Unifier, class Data>
+struct QueryRes
 {
-  TQueryRes() {}
-  TQueryRes(TermList t, Literal* l, Clause* c, Unifier unifier)
-  : term(t), literal(l), clause(c), unifier(std::move(unifier)) {}
-
-  TermList term;
-  Literal* literal;
-  Clause* clause;
   Unifier unifier;
+  Data const* data;
```

Comment:
Would it make sense for this to be a const reference instead? (Just curious.)

Reply:
I am also thinking of actually communicating this by value. (Where the value becomes a pointer if Data is a pointer.) This would mean it would only be efficient, if user's data is large, to require it to be heap allocated, before it's inserted into the index. Did you consider this option?

Reply:
I think it's best to keep it being a pointer here. So basically whenever we return a `QueryRes`, the data in this query result is stored somewhere in the leave of a substitution tree which is itself already heap allocated. So we don't have to access (dereference + copy) the data anywhere (not even to the heap) if we don't use it. Usually the data will consist of a term + some other things, which will be bigger in terms of memory than `Data const*`.

In terms of using a `Data const&` instead: Using templates together with references is quite a pain (implicit copies and other odd things that can occur rather unexpectedly). So this is why i opted for a `const*` instead here.

Reply:
Thanks for the explanations! I am glad to hear I am not the only one who finds working with references challenging ;)